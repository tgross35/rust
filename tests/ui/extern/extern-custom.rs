#![feature(abi_custom)]

unsafe extern "custom" {
    // Usual unsafe functions are allowed.
    fn extern_unsafe();

    // Trying to use `safe` is an error.
    safe fn extern_safe();
    //~^ ERROR `extern "custom"` functions cannot be marked `safe`
}

// By default, `custom` functions cannot be defined.
extern "custom" fn fn_def() {}
//~^ ERROR custom abi cannot be defined

// This should be rejected anyway, but check with closures (which inherit parent ABI).
extern "custom" fn fn_def_closure() {
    //~^ ERROR custom abi cannot be defined
    let x = || {};
    x();
}

// Naked functions can define functions with the `custom` ABI.
#[unsafe(naked)]
unsafe extern "custom" fn naked_fn_def() {
    core::arch::naked_asm!("nop")
}

// `custom` ABI functions must be marked `unsafe`.
#[unsafe(naked)]
extern "custom" fn naked_fn_def() {
    //~^ ERROR missing `unsafe`
    core::arch::naked_asm!("nop")
}

// Function parameters do nothing so are rejected.
#[unsafe(naked)]
unsafe extern "custom" fn naked_fn_def_params(a: u32) {
    //~^ ERROR no params allowed
    core::arch::naked_asm!("nop")
}

// Same for return types.
#[unsafe(naked)]
unsafe extern "custom" fn naked_fn_def_ret() -> u32 {
    //~^ ERROR no return types allowed
    core::arch::naked_asm!("nop")
}

// Check that passing as a function pointer works.
pub fn fn_ptr(_x: extern "unsafe" fn()) {}


// Check that usage as a `sym` in assembly works.
#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
unsafe fn use_as_sym() {
    unsafe {
        core::arch::asm!(
            "call {foo}",
            foo = sym foo,
        )
    }
}
