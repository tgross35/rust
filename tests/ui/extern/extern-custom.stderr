error[E0428]: the name `naked_fn_def` is defined multiple times
  --> $DIR/extern-custom.rs:31:1
   |
LL | unsafe extern "custom" fn naked_fn_def() {
   | ---------------------------------------- previous definition of the value `naked_fn_def` here
...
LL | extern "custom" fn naked_fn_def() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `naked_fn_def` redefined here
   |
   = note: `naked_fn_def` must be defined only once in the value namespace of this module

error[E0703]: invalid ABI: found `r#unsafe`
  --> $DIR/extern-custom.rs:51:26
   |
LL | pub fn fn_ptr(_x: extern "unsafe" fn()) {}
   |                          ^^^^^^^^ invalid ABI
   |
   = note: invoke `rustc --print=calling-conventions` for a full list of supported calling conventions

error[E0601]: `main` function not found in crate `extern_custom`
  --> $DIR/extern-custom.rs:63:2
   |
LL | }
   |  ^ consider adding a `main` function to `$DIR/extern-custom.rs`

error[E0570]: `"custom"` is not a supported ABI for the current target
  --> $DIR/extern-custom.rs:13:1
   |
LL | extern "custom" fn fn_def() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0570]: `"custom"` is not a supported ABI for the current target
  --> $DIR/extern-custom.rs:17:1
   |
LL | extern "custom" fn fn_def_closure() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0428, E0570, E0601, E0703.
For more information about an error, try `rustc --explain E0428`.
